70 implement blocks with different colors

KoLaR8 opened the pull request:
Solve #70 issue. Each block type has a different color.
![obraz](https://user-images.githubusercontent.com/76537707/122888374-1a735380-d342-11eb-9ed6-1cc0d704aff2.png)

TJB4epam commented on src/main/java/com/epam/prejap/tetris/block/Block.java:
```diff
@@ -1,11 +1,13 @@
     final int rows;
     final int cols;
-
+    final String color = ColorPicker.chooseRandomColor();
```
Is that it? Random colour each time? Was that mentioned in story card? My memory may mislead me but I thought it's colour assigned to a block?

KoLaR8 replied:
I was solving issue #70 , which says about new color for each block. After I actually write this code I saw that issue #12 is also about coloring blocks in game.

TJB4epam commented on src/main/java/com/epam/prejap/tetris/block/BlockFeed.java:
```diff
@@ -8,7 +8,8 @@
     private final Random rnd = new Random();
     private final List<Supplier<Block>> blocks = List.of(
-        OBlock::new
+            OBlock::new,
```
Not that I particularly mind, but why that additional indent?

TJB4epam commented on src/main/java/com/epam/prejap/tetris/block/TBlock.java:
```diff
@@ -0,0 +1,20 @@
+package com.epam.prejap.tetris.block;
```
Why is this reported as a new file that you're adding if it was added and merged by Grzegorz earlier?

TJB4epam commented on src/main/java/com/epam/prejap/tetris/colours/ColorPicker.java:
```diff
@@ -0,0 +1,63 @@
+import java.util.Random;
+
+/**
+ *  Purpose of this class is to menage different colors  behaviour provided from enum Colors.
```
Menage - French.
Manage - English. :-)

Also, there are two spaces between colours and behaviour.

TJB4epam replied:
Also - that's a rather vague name. I still don't know what this does, aside from "deals with colours", which I have from class name. Any better description?

TJB4epam commented on src/main/java/com/epam/prejap/tetris/colours/ColorPicker.java:
```diff
@@ -0,0 +1,63 @@
+ * @author Wojciech Kolarczyk
+ * @see Colors
+ */
+public class ColorPicker {
```
A curious case, here.

This class has NO data and all of it's API is `public static`. So, this is a tool class. However it works on an enum and accepts data from outside. 

Why then don't we move those methods to `enum Colours`?

If not, at least let's discourage instantiation of this class by adding a private default c-tor, with a comment, as per Effective Java.

TJB4epam commented on src/main/java/com/epam/prejap/tetris/colours/Colors.java:
```diff
@@ -0,0 +1,44 @@
+package com.epam.prejap.tetris.colours;
```
Not sure a part of UI of our game should be hidden in a package. Frankly, I fail to see the benefits this gives us.

I'd rather have this mixed with the UI package, or mixed into a package our UI hides / is used in.

Also, I have a feeling that if we combined this and the static method class ColorPicker we could deal away with 'getID' and 'getAsciCode' or however they are called.

Finally: ASCII, so asciiSomething, not asciSomething. Double i, even when in var name.

TJB4epam reviewed (changes requested):
Two large issues in a lovely approach to a nice issue.

1. Number of files I read here are seemingly new, while they aren't. This makes me confused about whether I'm reviewing YOUR code or whether I'm re-reviewing somebody else's code. Please straighten this out. `git rebase` should help.
2. I have a design doubt w.r.t. colours package. See my review remarks for details.

Thank you, requesting changes.

KoLaR8 commented:
Changes made:
1. moved content of ColorPicker class to enum Colors. 
2. Deleted ColorPicker class. 
3. Moved Colors to game package. 
4. Correct typos with 'asci' to "ascii", and with "menage" to "manage". 
5. Removed additional indent in BlockFeed.
6. Changed tests for OBlock

TJB4epam commented on pom.xml:
```diff
@@ -6,7 +6,7 @@
     <groupId>com.epam.prejap</groupId>
     <artifactId>tetris</artifactId>
-    <version>0.4</version>
+    <version>0.5</version>
```
TY!

TJB4epam commented on src/main/java/com/epam/prejap/tetris/block/Block.java:
```diff
@@ -39,4 +42,8 @@ public byte dotAt(int i, int j) {
         return image[i][j];
     }
 
+    public String  color(){
```
Colour is a String? :-O

TJB4epam replied:
OK, ASCII codes, OK.

TJB4epam commented on src/main/java/com/epam/prejap/tetris/block/BlockFeed.java:
```diff
@@ -8,8 +8,8 @@
-            OBlock::new,
-            TBlock::new
+        OBlock::new,
+        TBlock::new
```
TY

TJB4epam commented on src/main/java/com/epam/prejap/tetris/game/Colors.java:
```diff
@@ -0,0 +1,81 @@
+ * Each value contains ascii code and id.
+ * @author Wojciech Kolarczyk
+ */
+public enum Colors {
```
Plus for using ASCII colours, I don't think I mentioned that before but that's a nice move.

TJB4epam commented on src/test/java/com/epam/prejap/tetris/block/OBlockTest.java:
```diff
@@ -1,5 +1,6 @@
 package com.epam.prejap.tetris.block;
```
Once I merge Anna's changes these tests will have to be redone.

TJB4epam commented on src/main/java/com/epam/prejap/tetris/block/Block.java:
```diff
@@ -1,11 +1,13 @@
     final int rows;
     final int cols;
-
+    final String color = Colors.chooseRandomColor();
```
What I dislike about this is how we're using partial data out of the object, instead of relying on an object.

Why not have a Colours enum object here instead of a String?

TJB4epam commented on src/main/java/com/epam/prejap/tetris/game/Colors.java:
```diff
@@ -0,0 +1,81 @@
+        }
+        return color + "#" + Colors.RESET.asciiCode;
+    }
+}
```
If this enum had a decent toString method, we could work on it, like so:

WHITE.toString()

RED.toString()

RESET.toString()

in this manner, plenty of code that relies on a string from that class, could use type-safety, by using DIRECTLY some constant from this class. Like, WHITE or GREEN. Additional bonus comes in increased readability.

TJB4epam commented on src/main/java/com/epam/prejap/tetris/game/Colors.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+
+    public static String chooseRandomColor(){
```
Try moving away from statics in favour of methods on an object. Enum is one, so let's use it. You have the object with a state (ASCII code, id) and you can use this in methods - you already are.

TJB4epam commented on src/main/java/com/epam/prejap/tetris/game/Colors.java:
```diff
@@ -0,0 +1,81 @@
+            if(colors.asciiCode.equals(color)){
+                return colors.id;
+            }
+        }
```
Much better complexity here can be achieved with EnumMap. Instead of O(n) where n is number of colours we would have O(1).

TJB4epam reviewed (changes requested):
Take a look at the comments. I like the direction and this is better than previous version by far. We can lean on enums more and use them in a better, type-safer way.

KoLaR8 commented:
Changes made:
1. Rename Colors enum to Color.
2. Created toString method for Color.
3. Introduced Map for faster searching through Color values.
4. Replaced all string which were ascii codes with Color value.
5. Added messages to assertions in tests.
6. Changed tests in BlockShapesTest to fit my implementation. (Checking for color id, not for '1').
7. Added test for map.

TJB4epam commented:
This Wednesday we will focus on aggresively closing PRs. Just FYI.

KoLaR8 commented:
@TJB4epam 
updated to latest version with pom version to 11. Latest changes made me force to change tactic from generating random colour to block shape's defined colour, but it's much better and more tetris alike now. Please review as soon as possible, as this change is little bigger than others.

TJB4epam reviewed (changes requested):
See comments. Especially the one asking for complexity reduction - the longest.

AndriiVel commented:
Approved during the lecture. Version: 0.13

AndriiVel commented:
A follow-up issue will be created later (Complexity of feature, toStrings, enumMap)

marcin-ogorzalek-EPAM commented:
@KoLaR8 merge conflicts to solve

pkierat-epam commented:
Looks good. Merging.
